jujugui/static/gui/src/app/components/section-load-watcher/section-load-watcher.js

```diff
@@ -0,0 +1,129 @@
+SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
+General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+YUI.add('section-load-watcher', function(Y) {
+
+  juju.components.SectionLoadWatcher = React.createClass({
+
+    propTypes: {
+      // The eslint propType sorting requires that capital letters come
+      // before lower case.
+      EmptyComponent: React.PropTypes.func,
+      ErrorComponent: React.PropTypes.func,
+      children: React.PropTypes.oneOfType([
+        React.PropTypes.object,
+        React.PropTypes.array
+      ]),
+      timeout: React.PropTypes.number
+    },
+
+    getInitialState: function() {
+      this._childrenStatuses = new Map();
+
+      return {
+        renderEmpty: false,
+        renderError: false
+      };
+    },
+
+    _validStatuses: ['starting', 'ok', 'empty', 'error'],
+
+    /**
+      When a child status has been updated we need to loop through all of
+      the available statuses to see if they are all 'empty', 'error', or
+      if the timeout has been hit to then decide which component to show.
+
+      @method _checkChildrenStatuses
+    */
+    _checkChildrenStatuses: function() {
+      const total = React.Children.count(this.props.children);
+      let statuses = {};
+      this._childrenStatuses.forEach((status, ref) => {
+        if (!Number.isInteger(statuses[status])) {
+          statuses[status] = 0;
+        }
+        statuses[status] += 1;
+      });
+      if (statuses.empty === total) {
+        this.setState({renderEmpty: true});
+      }
+      if (statuses.error === total) {
+        this.setState({renderError: true});
+      }
+    },
+
+    /**
+      Sets the child status in the component and then checks all the statuses
+      to see if we should be rendering an error or empty component.
+
+      @method _setChildStatus
+      @param {String} ref The ref of the child.
+      @param {String} status The status of the child.
+    */
+    _setChildStatus: function(ref, status) {
+      if (this._validStatuses.indexOf(status) === -1) {
+        throw `Invalid status: "${status}" from ref: ${ref}`;
+      }
+      this._childrenStatuses.set(ref, status);
+      this._checkChildrenStatuses();
+    },
+
+    /**
+      Checks the state of the component to determine whether it should
+      render the empty, error, or children components.
+
+      @method _renderContent
+    */
+    _renderContent: function() {
+      if (this.state.renderEmpty) {
+        return <this.props.EmptyComponent />;
+      } else if (this.state.renderError) {
+        return <this.props.ErrorComponent />;
+      } else {
+        // Augment the children with the broadcastStatus method
+        const children = React.Children.map(this.props.children, child => {
```

Comment:
This clones the child components that were passed into the SLW and adds the `broadcastStatus` prop to it.